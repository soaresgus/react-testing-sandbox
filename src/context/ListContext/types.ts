interface IList {
  items: string[];
}

export interface IListContext extends IList {
  setItems(items: string[]): void;
}

export interface IListProvider {
  children: React.ReactNode;
}
