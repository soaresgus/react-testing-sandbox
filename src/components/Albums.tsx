import { useQuery } from '@tanstack/react-query';
import axios, { AxiosError } from 'axios';
import { useState } from 'react';

interface IAlbumData {
  userId: number;
  id: number;
  title: string;
}

export function Albums() {
  const [albumId, setAlbumId] = useState<number>(1);

  const { isLoading, data, isError, error } = useQuery({
    queryKey: [`album-${albumId}`],
    queryFn: () => getAlbumById(albumId),
  });

  async function getAlbumById(id: number) {
    const response = await axios.get(
      `https://jsonplaceholder.typicode.com/albums/${id}`
    );

    return response.data as IAlbumData;
  }

  function nextAlbum() {
    setAlbumId((state) => state + 1);
  }

  function previousAlbum() {
    setAlbumId((state) => (state > 1 ? state - 1 : 1));
  }

  if (isLoading) {
    return <span>Loading album...</span>;
  }

  if (isError) {
    return <span>{(error as AxiosError).message}</span>;
  }

  return (
    <div>
      <br />
      <br />
      <button onClick={previousAlbum} disabled={albumId <= 1}>
        Álbum anterior
      </button>
      <button onClick={nextAlbum}>Próximo Álbum</button>
      <h1>{data?.title}</h1>
    </div>
  );
}
