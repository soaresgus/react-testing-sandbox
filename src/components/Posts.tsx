import axios, { AxiosError } from 'axios';
import { useEffect, useState } from 'react';

interface IPostData {
  userId: number;
  id: number;
  title: string;
  body: string;
}

export function Posts() {
  const [posts, setPosts] = useState<IPostData[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<AxiosError>();

  async function getPosts() {
    const request = await axios.get(
      'https://jsonplaceholder.typicode.com/posts/1'
    );

    return request.data as IPostData;
  }

  useEffect(() => {
    getPosts()
      .then((data) => {
        setPosts([data]);
        setIsLoading(false);
      })
      .catch((error) => {
        setIsLoading(false);
        setError(error);
      });
  }, []);

  if (isLoading) {
    return <span>Loading...</span>;
  }

  return (
    <>
      {error ? (
        <span>{error.message}</span>
      ) : (
        posts.map((post) => (
          <div key={post.id}>
            <h1>{post.title}</h1>
            <span>{post.body}</span>
          </div>
        ))
      )}
    </>
  );
}
