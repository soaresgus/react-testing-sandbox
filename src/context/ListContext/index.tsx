import { createContext, useState } from 'react';
import { IListContext, IListProvider } from './types';

export const ListContext = createContext({} as IListContext);

export function ListProvider({ children }: IListProvider) {
  const [items, setItems] = useState<string[]>([]);

  return (
    <ListContext.Provider value={{ items, setItems }}>
      {children}
    </ListContext.Provider>
  );
}
