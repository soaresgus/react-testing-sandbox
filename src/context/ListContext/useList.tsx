import { useContext, useEffect } from 'react';
import { ListContext } from '.';

interface IUseListProps {
  initialItems?: string[];
}

export function useList({ initialItems }: IUseListProps) {
  const context = useContext(ListContext);

  useEffect(() => {
    if (initialItems) {
      context.setItems(initialItems);
    }
  }, []);

  return context;
}
