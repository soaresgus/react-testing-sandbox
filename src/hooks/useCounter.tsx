import { useState } from 'react';

export function useCounter() {
  const [count, setCount] = useState<number>(0);

  function increment() {
    setCount((state) => state + 1);
  }

  return { count, increment };
}
