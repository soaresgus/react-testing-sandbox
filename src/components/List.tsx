import { useState } from 'react';
import { useList } from '../context/ListContext/useList';

interface IListProps {
  initialItems: string[];
}

export function List({ initialItems }: IListProps) {
  const [newItem, setNewItem] = useState('');

  const { items, setItems } = useList({ initialItems });

  function addToList() {
    setTimeout(() => {
      setItems([...items, newItem]);
    }, 500);
  }

  function removeFromList(item: string) {
    setTimeout(() => {
      setItems(items.filter((value) => value != item));
    }, 500);
  }

  return (
    <>
      <label htmlFor="new-item">Novo item</label>
      <input
        id="new-item"
        type="text"
        placeholder="Novo item"
        value={newItem}
        onChange={(event) => setNewItem(event.target.value)}
      />
      <button onClick={addToList}>Adicionar</button>
      <ul>
        {items.map((item) => (
          <li key={item} data-testid="list-item">
            {item}
            <button onClick={() => removeFromList(item)}>Remover</button>
          </li>
        ))}
      </ul>
    </>
  );
}
