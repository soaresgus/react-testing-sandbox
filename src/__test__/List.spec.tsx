import {
  act,
  AllWrappers,
  render,
  renderHook,
  waitFor,
  waitForElementToBeRemoved,
} from './test-utils';
import userEvent from '@testing-library/user-event';
import { List } from '../components/List';
import { ListProvider } from '../context/ListContext';
import { useList } from '../context/ListContext/useList';
import { useCounter } from '../hooks/useCounter';

describe('List Component', () => {
  it('should render correctly', () => {
    const { container } = render(<List initialItems={['Julia']} />);

    expect(container).toMatchSnapshot();
  });

  it('should render list items', () => {
    const { getByText, queryByText, unmount } = render(
      <List initialItems={['Diego', 'Rodz', 'Mayk']} />,
      { wrapper: AllWrappers }
    );

    expect(getByText('Diego')).toBeInTheDocument();
    expect(getByText('Rodz')).toBeInTheDocument();
    expect(getByText('Mayk')).toBeInTheDocument();

    unmount();
    render(<List initialItems={['Julia']} />, { wrapper: ListProvider });
    expect(getByText('Julia')).toBeInTheDocument();
    expect(queryByText('Mayk')).not.toBeInTheDocument();
  });

  it('should be able to add new item to the list', async () => {
    const { getByText, getByPlaceholderText, getByLabelText, debug } = render(
      <List initialItems={[]} />
    );

    const inputElement = getByPlaceholderText('Novo item');
    const addButton = getByText('Adicionar');

    await userEvent.type(inputElement, 'Novo');
    await userEvent.click(addButton);

    await waitFor(() => {
      expect(getByText('Novo')).toBeInTheDocument();
    });
  });

  it('should be able to remove item from the list', async () => {
    const { getByText, getAllByText, debug } = render(
      <List initialItems={['Diego']} />
    );

    const removeButtons = getAllByText('Remover');

    await userEvent.click(removeButtons[0]);

    await waitForElementToBeRemoved(() => {
      return getByText('Diego');
    });
  });
});

describe('Counter Hook', () => {
  it('should increment count', () => {
    const { result } = renderHook(() => useCounter());

    act(() => {
      result.current.increment();
      result.current.increment();
      result.current.increment();
    });

    expect(result.current.count).toBe(3);
  });
});

describe('List Context', () => {
  it('shout set list items', () => {
    const { result } = renderHook(() => useList({ initialItems: ['Diego'] }), {
      wrapper: AllWrappers,
    });

    expect(result.current.items[0]).toBe('Diego');

    act(() => {
      result.current.setItems(['Marcelo']);
    });

    expect(result.current.items[0]).toBe('Marcelo');
  });
});
