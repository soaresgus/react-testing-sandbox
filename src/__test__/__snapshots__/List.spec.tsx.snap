// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component should render correctly 1`] = `
<div>
  <label
    for="new-item"
  >
    Novo item
  </label>
  <input
    id="new-item"
    placeholder="Novo item"
    type="text"
    value=""
  />
  <button>
    Adicionar
  </button>
  <ul>
    <li
      data-testid="list-item"
    >
      Julia
      <button>
        Remover
      </button>
    </li>
  </ul>
</div>
`;
